### PR TITLE
WIP: Validate metadata for custom resources

### LINKED ISSUE #59451: Namespaced CRD cannot used label selector
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

When listing a namespaced CRD with a valid label selector no results are returned

**What you expected to happen**:

Results returned

**How to reproduce it (as minimally and precisely as possible)**:

1. Create a namespaced CRD.
2. Create a resource with that CRD with labels
3. List the resources with a valid label selector

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 
`Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2", GitCommit:"5fa2db2bd46ac79e5e00a4e6ed24191080aa463b", GitTreeState:"clean", BuildDate:"2018-01-18T10:09:24Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2", GitCommit:"5fa2db2bd46ac79e5e00a4e6ed24191080aa463b", GitTreeState:"clean", BuildDate:"2018-01-18T09:42:01Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}`

**Test Resources**:

```
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: volumes.sandwichcloud.com
spec:
  group: sandwichcloud.com
  names:
    kind: Volume
    listKind: VolumeList
    plural: volumes
    singular: volume
  scope: Namespaced
  version: v1alpha1
---
apiVersion: sandwichcloud.com/v1alpha1
kind: Volume
metadata:
  finalizers:
  - delete.sandwichcloud.com
  labels:
    sandwichcloud.com/attached_to: null
    sandwichcloud.com/id: 9f52f5b2-559a-475e-95f1-c7b8f0067116
    sandwichcloud.com/name: vol1
    sandwichcloud.com/region: e7444046-fc8b-4c40-81e2-357a5c80168a
    sandwichcloud.com/zone: 88f97cb1-6375-4a13-b9ea-bfd013a33bbd
  name: 9f52f5b2-559a-475e-95f1-c7b8f0067116
spec:
  backingId: 892a404d-60f8-4263-9c5a-5eb45906eb06
  clonedFrom: null
  size: 10
status:
  errorMessage: ""
  state: Created
  task:
    kwargs: {}
    name: null
```

Run `kubectl get volumes -l sandwichcloud.com/name=vol1` and it will return no results.

```
I0206 19:26:04.343321   11513 loader.go:357] Config loaded from file kubeconfig.yml
I0206 19:26:04.365620   11513 round_trippers.go:417] curl -k -v -XGET  -H "Accept: application/json" -H "User-Agent: kubectl/v1.9.2 (linux/amd64) kubernetes/5fa2db2" http://localhost:8081/apis/sandwichcloud.com/v1alpha1/namespaces/e7a3017c-36fb-466f-97f6-d7079b31e5aa/volumes?labelSelector=sandwichcloud.com%2Fname%3Dvol1&limit=500
I0206 19:26:04.375805   11513 round_trippers.go:436] GET http://localhost:8081/apis/sandwichcloud.com/v1alpha1/namespaces/e7a3017c-36fb-466f-97f6-d7079b31e5aa/volumes?labelSelector=sandwichcloud.com%2Fname%3Dvol1&limit=500 200 OK in 7 milliseconds
I0206 19:26:04.377152   11513 round_trippers.go:442] Response Headers:
I0206 19:26:04.377662   11513 round_trippers.go:445]     Content-Type: application/json
I0206 19:26:04.378014   11513 round_trippers.go:445]     Date: Wed, 07 Feb 2018 01:26:04 GMT
I0206 19:26:04.378400   11513 round_trippers.go:445]     Content-Length: 229
I0206 19:26:04.379130   11513 request.go:873] Response Body: {"apiVersion":"sandwichcloud.com/v1alpha1","items":[],"kind":"VolumeList","metadata":{"continue":"","resourceVersion":"2574","selfLink":"/apis/sandwichcloud.com/v1alpha1/namespaces/e7a3017c-36fb-466f-97f6-d7079b31e5aa/volumes"}}
No resources found.
```

### PR DESCRIPTION
**Pull request by nikhita**
Validates the following by converting the metadata subtree of custom resources from `Unstructured` to a metadata struct and back:

1. null/nil values for labels and annotations are marshalled as empty strings `""`.
2. Validates metadata to make sure there are no unknown fields.
3. Validates invalid values (numbers, booleans, structured values, etc) for labels, annotations, etc.

Fixes #59451

TODO:

- [ ] Decide how to proceed with existing bad data

**Release note**:

(This probably deserves a release note but will update after the TODO is done)

```release-note
NONE
```

/cc sttts liggitt deads2k ash2k jennybuckley

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/validation_test.go [hunk 1]
```diff
@@ -30,61 +30,6 @@ import (
 	"k8s.io/apiextensions-apiserver/test/integration/testserver"
 )
 
-func TestForProperValidationErrors(t *testing.T) {
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/validation_test.go]: This is a pure move.

**Comment by liggitt** [comment]
Is there a way to use the `unstructured<->typed` converter to convert the metadata subtree to a real Metadata struct, then back, in PrepareForCreate and PrepareForUpdate?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+func ValidateObjectMetaFields(objectMeta map[string]interface{}) error {
+	emptyObjectMeta := metav1.ObjectMeta{}
+
+	// detect unknown fields in metadata
+	v := reflect.ValueOf(emptyObjectMeta)
+	totalFields := v.NumField()
+	objectMetaFieldNames := make(map[string]bool, totalFields)
+
+	for i := 0; i < totalFields; i++ {
+		jsonName := strings.Split(v.Type().Field(i).Tag.Get("json"), ",")[0]
```
Comment (wgliang) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go]: This index approach should be added to determine the safety of the approach, or there may be a risk of panic.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	}
+	if typeAccessor.GetKind() != a.kind.Kind {
+		return field.ErrorList{field.Invalid(field.NewPath("kind"), typeAccessor.GetKind(), fmt.Sprintf("must be %v", a.kind.Kind))}
+	}
+	if typeAccessor.GetAPIVersion() != a.kind.Group+"/"+a.kind.Version {
+		return field.ErrorList{field.Invalid(field.NewPath("apiVersion"), typeAccessor.GetAPIVersion(), fmt.Sprintf("must be %v", a.kind.Group+"/"+a.kind.Version))}
+	}
+
+	customResourceObject, ok := obj.(*unstructured.Unstructured)
+	// this will never happen.
```
Comment (wgliang) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go]: I suggest to remove this comment.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	}
+
+	return validation.ValidateObjectMetaAccessorUpdate(objAccessor, oldAccessor, field.NewPath("metadata"))
+}
+
+// ValidateObjectMetaFields validates the metadata to make sure it does not contain unknown fields and invalid values.
+func ValidateObjectMetaFields(objectMeta map[string]interface{}) error {
+	emptyObjectMeta := metav1.ObjectMeta{}
+
+	// detect unknown fields in metadata
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go]: this isn't the same behavior as native resources... we should drop unknown fields, not error on them (otherwise our API server is not forward compatible with ignoring new metadata fields)

**Comment by ash2k** [comment]
> Is there a way to use the unstructured<->typed converter to convert the metadata subtree to a real Metadata struct, then back, in PrepareForCreate and PrepareForUpdate?

I think that would be the best approach. Can be done using `k8s.io/apimachinery/pkg/runtime/DefaultUnstructuredConverter.(From|To)Unstructured()`.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go (hunk 3 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go]: or better move it inside the if clause.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go [hunk 5]
```diff
@@ -63,9 +58,35 @@ func (a customResourceDefinitionStorageStrategy) NamespaceScoped() bool {
...
 func (customResourceDefinitionStorageStrategy) PrepareForUpdate(ctx genericapirequest.Context, obj, old runtime.Object) {
+	customResourceObject, _ := obj.(*unstructured.Unstructured)
+	metadata := customResourceObject.UnstructuredContent()["metadata"]
+
+	// convert metadata subtree to metadata struct and back
+	// this is done to make sure we drop unknown fields from metadata
+	// if there are errors, we don't convert and fail in ValidateCreate and ValidateUpdate
+	in := metadata.(map[string]interface{})
+	out := &metav1.ObjectMeta{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(in, out); err == nil {
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go]: return early rather than nesting

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go [hunk 6]
```diff
@@ -63,9 +58,35 @@ func (a customResourceDefinitionStorageStrategy) NamespaceScoped() bool {
...
+		if convertedMetadata, err := runtime.DefaultUnstructuredConverter.ToUnstructured(out); err == nil {
+			customResourceObject.UnstructuredContent()["metadata"] = convertedMetadata
+		}
+	}
 }
 
 func (customResourceDefinitionStorageStrategy) PrepareForUpdate(ctx genericapirequest.Context, obj, old runtime.Object) {
+	customResourceObject, _ := obj.(*unstructured.Unstructured)
+	metadata := customResourceObject.UnstructuredContent()["metadata"]
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go]: `, ok` and return early if missing

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go [hunk 7]
```diff
@@ -115,73 +136,3 @@ func (a customResourceDefinitionStorageStrategy) MatchCustomResourceDefinitionSt
 		GetAttrs: a.GetAttrs,
 	}
 }
-
-type customResourceValidator struct {
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go]: hard to follow changes combined with the move... can you keep it in place or separate the straight move into a separate commit?

Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go]: can keep it in place. the move is in the subresources PR (adds lots of things there).

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go [hunk 8]
```diff
@@ -63,9 +58,35 @@ func (a customResourceDefinitionStorageStrategy) NamespaceScoped() bool {
 }
 
 func (customResourceDefinitionStorageStrategy) PrepareForCreate(ctx genericapirequest.Context, obj runtime.Object) {
+	customResourceObject, _ := obj.(*unstructured.Unstructured)
+	metadata := customResourceObject.UnstructuredContent()["metadata"]
+
+	// convert metadata subtree to metadata struct and back
+	// this is done to make sure we drop unknown fields from metadata
+	// if there are errors, we don't convert and fail in ValidateCreate and ValidateUpdate
+	in := metadata.(map[string]interface{})
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go]: `, ok` and short-circuit... hard casting here can panic

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go]: probably want a helper that does this coercion so we can call it from PrepareForCreate, PrepareForUpdate, and validation functions

that would allow unit testing various scenarios as well

I had a WIP doing something similar in https://github.com/liggitt/kubernetes/commits/crd-coercion in case it's useful.

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go]: also, based on the issues I was hitting, I don't think fromUnstructured does the nil -> "" conversion we wanted

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+	}
+	_, err = runtime.DefaultUnstructuredConverter.ToUnstructured(out)
+	if err != nil {
+		return field.ErrorList{field.Invalid(field.NewPath("metadata"), nil, err.Error())}
+	}
+
+	return validation.ValidateObjectMetaAccessor(accessor, a.namespaceScoped, validation.NameIsDNSSubdomain, field.NewPath("metadata"))
+}
+
+func (a customResourceValidator) ValidateUpdate(ctx genericapirequest.Context, obj, old runtime.Object) field.ErrorList {
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go]: need to think about what we do here, since we have the possibility of persisted mis-structured data.

ratcheting validation seems like a good start (if the old object coerces without error, require the new object to as well).

dropping fields with structural errors (along with unknown fields) is another possibility (it's likely what we would try to do to fix the issues), but is not as simple to do.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go (hunk 8 above)
Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go]: `structuredMetadata` for the `ValidateMetadata` test:

```
&ObjectMeta{Name:foo,GenerateName:,Namespace:not-the-default,SelfLink:/apis/mygroup.example.com/v1beta1/namespaces/not-the-default/noxus/foo,UID:169ff77e-158a-11e8-9dd2-54e1ad6c2d05,ResourceVersion:6,Generation:0,CreationTimestamp:2018-02-19 21:02:26 +0530 IST,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{qux: ,},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,}
```

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go (hunk 9 above)
Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go]: this is still open.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go (hunk 8 above)
Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go]: > also, based on the issues I was hitting, I don't think fromUnstructured does the nil -> "" conversion we "wanted"

Also, from this test data - https://github.com/liggitt/kubernetes/commit/51965f9e381e9588d918ad1281bf94b1bec65c58#diff-1b3f35816cc25f4fd5cd83f1c18c013fR327:

The unstructured object is:
```
map[bb:<nil> bc:<nil> bd:<nil> ba:map[aa:<nil> ab:<nil> ac:<nil>]]
```

and both `newObj` and `unmarshalledObj` are:

```
{"ba":{},"bb":"","bc":null,"bd":null}
```

which seems right to me. @liggitt thoughts?

**Comment by nikhita** [comment]
Closing in favor of pruning: https://github.com/kubernetes/kubernetes/pull/63604.

Also, the pruning PR exists behind a feature gate so it takes care of the following TODO:

> Decide how to proceed with existing bad data

/close
